core/src/main/scala/org/apache/spark/sql/execution/joins/LocalJoin.scala

```diff
@@ -161,57 +198,81 @@ case class LocalJoin(leftKeys: Seq[Expression],
-        scan.input = rowIterator; scan
-    }.getOrElse(throw new IllegalStateException(
-      s"Failed to find replicated table for LocalJoin in $buildPlan"))
-    val rdd = rowTableScan.dataRDD
-    assert(rdd.getNumPartitions == 1)
-    val rowTableRDD = ctx.addReferenceObj("rowTableRDD", rdd)
-    val rowTablePart = ctx.addReferenceObj("singlePartition", rdd.partitions(0))
     // using the expression IDs is enough to ensure uniqueness
+    val buildCodeGen = buildPlan.asInstanceOf[CodegenSupport]
+    val rdds = buildCodegenRDDs
     val exprIds = buildPlan.output.map(_.exprId.id).toArray
     val cacheKeyTerm = ctx.addReferenceObj("cacheKey",
-      new CacheKey(exprIds, rdd.id))
+      new CacheKey(exprIds, rdds.head.id))
 
     // generate local variables for HashMap data array and mask
     mapDataTerm = ctx.freshName("mapData")
     maskTerm = ctx.freshName("hashMapMask")
     keyIsUniqueTerm = ctx.freshName("keyIsUnique")
     numRowsTerm = ctx.freshName("numRows")
+
     // generate the map accessor to generate key/value class
     // and get map access methods
     val session = sqlContext.sparkSession.asInstanceOf[SnappySession]
     mapAccessor = ObjectHashMapAccessor(session, ctx, buildSideKeys,
       buildPlan.output, "LocalMap", hashMapTerm, mapDataTerm, maskTerm,
       multiMap = true, this, this.parent, buildPlan)
 
-    val entryClass = mapAccessor.getClassName
-    val numKeyColumns = buildSideKeys.length
+    val buildRDDs = ctx.addReferenceObj("buildRDDs", rdds.toArray,
+      s"${classOf[RDD[_]].getName}[]")
+    val buildParts = rdds.map(_.partitions)
+    val partitionClass = classOf[Partition].getName
+    val buildPartsVar = ctx.addReferenceObj("buildParts", buildParts.toArray,
+      s"$partitionClass[][]")
+    val allIterators = ctx.freshName("allIterators")
+    val indexVar = ctx.freshName("index")
     val contextName = ctx.freshName("context")
     val taskContextClass = classOf[TaskContext].getName
     ctx.addMutableState(taskContextClass, contextName,
       s"this.$contextName = $taskContextClass.get();")
-    ctx.addMutableState("scala.collection.Iterator", rowIterator,
-      s"this.$rowIterator = $rowTableRDD.iterator($rowTablePart, $contextName);")
-    ctx.addNewFunction(getOrCreateMap,
+
+
+    // switch inputs to use the buildPlan RDD iterators
+    ctx.addMutableState("scala.collection.Iterator[]", allIterators,
       s"""
+         |$allIterators = inputs;
+         |inputs = new scala.collection.Iterator[$buildRDDs.length];
+         |$taskContextClass $contextName = $taskContextClass.get();
+         |for (int $indexVar = 0; $indexVar < $buildRDDs.length; $indexVar++) {
+         |  $partitionClass[] parts = $buildPartsVar[$indexVar];
+         |  // check for replicate table
+         |  if (parts.length == 1) {
+         |    inputs[$indexVar] = $buildRDDs[$indexVar].iterator(
+         |      parts[0], $contextName);
+         |  } else {
```

Comment:
Since you have the "replicatedTableJoin" flag, it may be better to create two different snippets based on that and not depend on parts.length (rather parts.length can have an assertion for that case)